crates/relayer-types/src/applications/transfer/msgs/transfer.rs

```diff
@@ -41,6 +41,7 @@ pub struct MsgTransfer<C = Coin> {
     pub timeout_timestamp: Timestamp,
     /// optional memo
     pub memo: Option<String>,
+    pub tokens: Vec<C>,
```

Comment:
Let's add a doc comment here.

I also wonder if we can't just keep this `tokens` field only instead of having both `token` and `tokens`?

Reply:
If I'm not mistaken, a channel with version `ics20-1` will require the `token` field since the previous proto didn't have the `tokens` field https://github.com/cosmos/ibc-go/blob/v8.3.1/proto/ibc/applications/transfer/v1/tx.proto#L51.

Reply:
Sure, but this is a domain type, so we can instead opt to only keep a single field here (ie. `tokens`), which would only ever contain a single `Coin` for v1 and potentially multiple for v2. That is, as long as we can still tell somehow which version we are dealing with, maybe from the context?

Or are we using the fact that `token` is `Some` or `None` to figure out which version it is?

Reply:
Oh this is a good point. I tried to keep the domain type as close to the proto struct as possible since the crate could be used by someone else.

We don't use the fact that `token` is `Some` or `None` to determine the version, but the implementation `From<MsgTransfer> for RawMsgTransfer` might be tricky.
https://github.com/informalsystems/hermes/blob/d05bc5a468904816b669fcc52fe3db64a4315a24/crates/relayer-types/src/applications/transfer/msgs/transfer.rs#L93
If there is only 1 token in the domain type's `tokens`, we would need to find a way to determine if we are building a v1 `RawMsgTransfer` which will have the `token: Some(token)` or the v2 with `tokens: vec!(token)`. But maybe ibc-go has some type of conversion since there is already a conversion for the PacketData and this isn't an issue, https://github.com/cosmos/ibc-go/blob/main/modules/apps/transfer/internal/convert/convert.go

Reply:
How about making the `MsgTransfer` domain type an enum, with one variant for v1 and one for v2?

Reply:
I updated `MsgTransfer` to be an enum in this commit https://github.com/informalsystems/hermes/pull/4008/commits/8ac128e63f7af43e84857c098a49017a69c18f7f